### ISSUE TITLE
Support PHI-Canto by adding meta-genotypes to the Canto database schema

### ISSUE THREAD

**Issue by kimrutherford** [pombase/canto#1442]
After discussion with @jseager7 and @barnabynorman we have a plan for storing the host+pathogen / meta-genotype in the database.  We plan to store the meta-genotypes in the existing genotype table but with a flag that says it's not a normal genotype.  The two parts of the meta-genotype would be stored as separate rows in the genotype table and we'll add connecting table to join to genotype rows to the meta-genotype rows.

This plan allows us to model the combined host-pathogen genotype with modest changes to the database schema and to the code.

 - add a `is_meta_genotype` boolean column to genotype table
 - add a `host_organism` column to handle the case where there are no host alleles in meta-genotype
 - add a `genotype_relations` table to connect the meta-genotype to it's parts

**Comment by kimrutherford** [comment]
> add a is_meta_genotype boolean column to genotype table
> add a host_organism column to handle the case where there are no host alleles in meta-genotype
> add a genotype_relations table to connect the meta-genotype to it's parts

After a bit more thought I think:
 - maybe we can infer `is_meta_genotype` so it can be a method rather than a database column
 - the `genotype_relations` should instead be called something like `metagenotype_part` 

The `metagenotype_part` table shoud look something like:
 - `metagenotype_id` (references the genotype table)
 - `is_host_part` (true if and only if this is the host part of the metagenotype)
 - `organism_id` (references the host or pathogen organism - can't be null)
 - `genotype_id` (references the genotype table - can be null for metagenomes with no host alleles, where `part_type` is "host")

A metagenotype with no host alleles would have:
 1. a `genotype` table entry with no alleles attached - this is the genotype row that will have annotation attached
 2. a `genotype` table entry for the pathogen genotype, which references the pathogen alleles via the `allele_genotype` table
 3. a `metagenotype_part` table entry:
    - `metagenotype_id` references the genotype table entry that has no alleles ("1" above)
    - `genotype_id` references the pathogen genotype ("2" above)
    - `is_host_part` will be `true`
    - `organism_id` will reference the host organism (a row in the organism table)

A metagenotype with host and pathogen alleles would have:
 1. a `genotype` table entry with no alleles attached - this is the genotype row that will have annotation attached
 2. a `genotype` table entry for the **pathogen** genotype, which references the pathogen alleles via the `allele_genotype` table
 3. a `genotype` table entry for the **host** genotype, which references the host alleles via the `allele_genotype` table
 4. a **pathogen** `metagenotype_part` table entry where:
    - `metagenotype_id` references the genotype table entry that has no alleles ("1" above)
    - `genotype_id` references the **pathogen** genotype ("2" above)
    - `is_host_part` will be `false`
    - `organism_id` will reference the **pathogen** organism
 5. a **host** `metagenotype_part` table entry where:
    - `metagenotype_id` references the genotype table entry that has no alleles ("1" above)
    - `genotype_id` references the **host** genotype ("3" above)
    - `is_host_part` will be `true`
    - `organism_id` will reference the **host** organism

This second example involves 5 table rows but all annotation will be attached the row in "1" above.  Because that's a genotype table entry most of the existing annotation code should need few changes.

The code that displays genotype/phenotype annotation in tables and anywhere else the shows genotypes and allele details will need more changes.

**Comment by jseager7** [comment]
I've tried to make a model of this, but I'm having some trouble figuring out how the genotypes relate to the `metagenotype_part` and the `organism`. Could you let me know if these models are on the right track?

**Pathogen-only**

Here's what I have for the case of a meta-genotype with a pathogen genotype but no host genotype:

![meta-genotype_host_only](https://user-images.githubusercontent.com/37659591/41405044-b64ce084-6fc0-11e8-950d-b1d5325dc804.png)

In your explanation you stated that

> metagenotype_id references the genotype table entry that has no alleles ("1" above)

and also that

> is_host_part will be true

So I'm wondering if there's an existing relation to the (no-allele) genotype and an organism entity: that is, which of the labelled relations in the above diagram is correct (1 or 2), or are neither of them accurate?

**Pathogen and host**

I've got a similar issue with the case where there is a pathogen genotype and a host genotype:

![meta-genotype_host_pathogen](https://user-images.githubusercontent.com/37659591/41405101-db563d58-6fc0-11e8-84cd-9a092206f32f.png)

Again, do any of the labelled relations apply? If so, which ones?

**Comment by kimrutherford** [comment]
Thanks for making diagrams.  That's really useful.

> Again, do any of the labelled relations (1, 2, 3, or 4) apply? If so, which ones?

I don't think those relations will be needed because we can always get from the genotype or meta-genotype to the organism via the metagenotype table.

But your diagrams have made me rethink how the organisms and genotypes should be connected.  It might be more sensible to reference the organism from the genotype and have a host genotype object/row even in the case where there are no alleles.  That makes sense biologically because the host always has a genotype.  It might just be that the genotype is the wild type.

I've hacked your diagram to give this:

![metagenome-shema-diagram-1](https://user-images.githubusercontent.com/90474/41452603-c56a3088-70c6-11e8-8140-f8a8ee5b907f.png)

**Comment by kimrutherford** [comment]
> I've hacked your diagram to give this:

After further thought I think we can remove the "is_host_part" fields because we can always look at the organism of the genotype to work out if it is host or pathogen.  And if we remove that field it might make sense to remove "metagenotype_part" too and then add a "metagenotype_id" field to genotype.  So the diagram would collapse to:

![metagenome-shema-diagram-2](https://user-images.githubusercontent.com/90474/41507483-45dc2d9a-7287-11e8-8b12-91f2c3f60df6.png)

I've also created a branch for trying this out: https://github.com/pombase/canto/tree/tickets/ticket-1442

**Comment by kimrutherford** [comment]
Yesterday James suggested adding a "metagenotype" table rather than dual purposing the genotype table to hold meta-genotypes too.

I had thought that adding a new table would be the option that takes more work.  After James' suggestion I had another look at the Perl code and now I'm not sure.  I think both options are would be painful but each option has different pain.  So now I don't know.  :-)

**Comment by jseager7** [comment]
For reference, this is what the the object model might look like if we added a meta-genotype table:

![metagenotype_model_alt](https://user-images.githubusercontent.com/37659591/41587658-a187b5a8-73a7-11e8-979d-873f8e418fd8.png)

Personally I would go for the option that best fits the actual domain model (and by that I mean the way _we_ think about these concepts, since a meta-genotype doesn't exist in the real world anyway), because I think the simplest description of a meta-genotype is that it is a combination (or composition) of two genotypes: a host genotype and a pathogen genotype.

It seems like with the other option (if I'm interpreting it correctly) we include a third genotype that takes the _role_ of composing the other two genotypes, but the existence of this third generic genotype isn't really obvious from the domain model. I'm a bit concerned that combining all this information into the genotype table (by not having a separate entity modelling a meta-genotype) will mean that the meta-genotype is heavily inferred, and somewhat ambiguous.

I'd like the database model to match the domain model as closely as possible, because I assume any inconsistencies or ambiguities there are bound to bleed into the Object Relational Mapper and the rest of the application. But again, it depends what the different types of pain are with these two choices. Equally, we probably don't want the database model to be a completely different type of abstraction to what the front-end requires, since that just creates more mapping work.

**Comment by kimrutherford** [comment]
> For reference, this is what the the object model might look like if we added a meta-genotype table:

I've been working on this today and I've now implemented what you have in the diagram.  I've also added a new web service end-point for storing metagenotypes.  I still need to add some more tests and make a AngularJS service that uses the new end-point.  I plan to do that tomorrow.

The changes so far are in this branch: https://github.com/pombase/canto/tree/tickets/ticket-1442
